keps/sig-architecture/3136-beta-apis-off-by-default/README.md

```diff
@@ -0,0 +1,293 @@
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+New beta APIs will not be enabled in clusters by default.
+Existing beta APIs and new versions of existing beta APIs, will continue to be enabled by default: 
+if v1beta.some.group is currently enabled by default and we create v1beta2.some.group, v1beta2.some.group will still be enabled by default.
+
+## Motivation
+
+Beta APIs are not considered stable and reliance upon APIs in this state leads to exposure to bugs,
+guaranteed migration pain for users when the APIs move to stable, and the risk that dependencies will
+grow around unfinished APIs.
+Enabling beta APIs by default, exacerbates these problems by making them on in nearly every cluster.
+We observed these problems as we removed long-standing beta APIs and the PRR survey tells us that over
+90% of production clusters leave these APIs enabled.
+Unsuitability for production use is documented at https://kubernetes.io/docs/reference/using-api/#api-versioning 
+("The software is not recommended for production uses"), but defaulting on means they are present in nearly every 
+production cluster.
+By disabling beta APIs by default, a cluster-admin can opt-in for specific APIs without having every
+incomplete API present in the cluster.
+
+### Goals
+
+1. Disable new beta APIs by default.
+2. Continue enabling existing beta APIs and new version of existing beta APIs by default:
+   if v1beta.some.group is currently enabled by default and we create v1beta2.some.group, v1beta2.some.group will still be enabled by default.
+3. Allow enabling specific resources in beta.  Enable coolnewjobtype.v1beta1.batch.k8s.io without enabling other-neat-job.v1beta1.batch.k8s.io
+
+### Non-Goals
+
+1. Change featuregate defaults.
+
+## Proposal
+
+New beta APIs will be placed into the `DisableVersions` stanza instead of the `EnableVersions` stanza (see [DefaultAPIResourceConfigSource](https://github.com/kubernetes/kubernetes/blob/0669da445fa8c1ae07c15c0827f0e83da11cbe58/pkg/controlplane/instance.go#L643)).
+The `--runtime-config` flag will be extended to allow `group/version/resource=true`, to enable specific resources.
+To enable a beta API, a cluster-admin will have to add the appropriate `--runtime-config` flags.
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a cluster-admin I want to enable the coolnewjobtype.v1beta1.batch.k8s.io API in my cluster.
+
+To do this I call `kube-apiserver --runtime-config=batch.k8s.io/v1beta1/coolnewjobtype`.
+
+#### Story 2
+
+As a cluster-admin I want to enable all beta APIs as in past releases.
+
+To do this I call `kube-apiserver --runtime-config=api/beta=true`.
+This already exists and will continue to function.
+
+
+### Notes/Constraints/Caveats (Optional)
+
+### Risks and Mitigations
+
```

Comment:
Another missing risk is that large features become essentially impossible to add incrementally, because the feedback loop is too long. I doubt either SSA or APF would have been able to make progress if this were the rule.

Reply:
good point

Reply:
> Another missing risk is that large features become essentially impossible to add incrementally, because the feedback loop is too long. I doubt either SSA or APF would have been able to make progress if this were the rule.

APF would definitely have gotten feedback.

SSA did go through some API evolution, though how much was based on "on-by-default" versus, "I want to try this" isn't immediately clear to me.  We had some pretty knowledgeable users providing feedback.